File: leetcode/easy/28_implement_strStr/mac.swift
// https://leetcode.com/problems/implement-strstr/
// Runtime: 12 ms, faster than 79.07% of Swift online submissions for Implement strStr().
// Memory Usage: 20.4 MB, less than 100.00% of Swift online submissions for Implement strStr().

class Solution {
    func strStr(_ haystack: String, _ needle: String) -> Int {
        if needle.isEmpty {
            return 0
        } else if haystack.count < needle.count {
            return -1
        }
        
        let count = haystack.count - needle.count
        for i in 0...count {
            let startIndex = haystack.index(haystack.startIndex, offsetBy: i)
            let endIndex = haystack.index(startIndex, offsetBy: needle.count)
            if String(haystack[startIndex..<endIndex]) == needle {
                return i
            }
        }

        return -1
    }
}

File: leetcode/easy/28_implement_strStr/mac.js
// https://leetcode.com/problems/implement-strstr/
// Runtime: 52 ms, faster than 90.40% of JavaScript online submissions for Implement strStr().
// Memory Usage: 34 MB, less than 67.86% of JavaScript online submissions for Implement strStr().

/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function(haystack, needle) {
    const main = {
        init : function(e, n) {
            return e.indexOf(n);
        }
    }
    
    return main.init(haystack, needle);
};